packages/components/src/components/permissions/models.ts

```diff
@@ -0,0 +1,220 @@
+}
+
+export class SecurityPolicy extends Record({
+    assignments: List<SecurityAssignment>(),
+    assignmentsByRole: Map<string, List<SecurityAssignment>>(),
+    modified: undefined,
+    resourceId: undefined,
+    relevantRoles: List<string>()
+}) {
+    assignments: List<SecurityAssignment>;
+    assignmentsByRole: Map<string, List<SecurityAssignment>>;
+    modified: string;
+    resourceId: string;
+    relevantRoles: List<String>;
+
+    constructor(values?: {[key:string]: any}) {
+        super(values);
+    }
+
+    static create(raw: any): SecurityPolicy {
+        let assignments = List<SecurityAssignment>();
+        if (raw.policy && raw.policy.assignments) {
+            assignments = List<SecurityAssignment>(raw.policy.assignments.map((assignment) => new SecurityAssignment(assignment)));
+        }
+
+        let relevantRoles = List<string>();
+        if (raw.relevantRoles) {
+            relevantRoles = List<string>(raw.relevantRoles);
+        }
+
+        return new SecurityPolicy({
+            ...raw.policy,
+            assignments,
+            assignmentsByRole: SecurityPolicy.getAssignmentsByRole(assignments),
+            relevantRoles
+        });
+    }
+
+    static getAssignmentsByRole(assignments: List<SecurityAssignment>): Map<string, List<SecurityAssignment>> {
+        let assignmentsByRole = Map<string, List<SecurityAssignment>>();
+
+        assignments.map((assignment) => {
+            if (!assignmentsByRole.has(assignment.role)) {
+                assignmentsByRole = assignmentsByRole.set(assignment.role, List<SecurityAssignment>());
+            }
+
+            const principals = assignmentsByRole.get(assignment.role);
+            assignmentsByRole = assignmentsByRole.set(assignment.role, principals.push(assignment));
+        });
+
+        return assignmentsByRole;
+    }
+
+    static removeAssignment(policy: SecurityPolicy, userId: number, role: SecurityRole): SecurityPolicy {
+        const assignments = policy.assignments.filter((assignment) => !(assignment.role === role.uniqueName && assignment.userId === userId)).toList();
+
+        return policy.merge({
+            assignments,
+            assignmentsByRole: SecurityPolicy.getAssignmentsByRole(assignments)
+        }) as SecurityPolicy;
+    }
+
+    static addAssignment(policy: SecurityPolicy, principal: Principal, role: SecurityRole): SecurityPolicy {
+        const assignments = policy.assignments.push(new SecurityAssignment(
+            Object.assign({role: role.uniqueName}, principal.toJS())
+        ));
+
+        return policy.merge({
+            assignments,
+            assignmentsByRole: SecurityPolicy.getAssignmentsByRole(assignments)
+        }) as SecurityPolicy;
+    }
+
+    static updateAssignmentsData(policy: SecurityPolicy, principalsById: Map<number, Principal>): SecurityPolicy {
+        const assignments = policy.assignments.map((assignment) => {
+            const principal = principalsById.get(assignment.userId);
```

Comment:
Any need to check for principal being undefined here?

Reply:
probably not, but simple enough to add